File: examples/orders-service-py/orders/services/order_service.py
"""The heart of the flow: a request becomes a stored, confirmed order."""

from __future__ import annotations

from ..domain.order import Order, OrderRequest
from ..notifications.email_service import EmailService
from ..pricing.pricing_service import PricingService
from ..repository.order_repository import OrderRepository
from ..validation.order_validator import validate_order_request

_sequence = 0


class OrderService:
    """Validates, prices, stores, and confirms new orders."""

    def __init__(self, pricing: PricingService, repository: OrderRepository, email: EmailService) -> None:
        self._pricing = pricing
        self._repository = repository
        self._email = email

    def place_order(self, request: OrderRequest) -> Order:
        """Validate, price, store, and confirm a new order."""
        validate_order_request(request)
        money = self._pricing.price(request)
        order = self._assemble(request, money)
        self._repository.save(order)
        self._email.send_order_confirmation(order)
        return order

    def get_order(self, order_id: str) -> Order | None:
        """Look up an order that was placed earlier."""
        return self._repository.find_by_id(order_id)

    def _assemble(self, request: OrderRequest, money: dict[str, int]) -> Order:
        """Combine the request with its computed money fields into a final order."""
        return Order(
            id=self._next_id(),
            customer_id=request.customer_id,
            lines=request.lines,
            subtotal_cents=money["subtotal_cents"],
            discount_cents=money["discount_cents"],
            tax_cents=money["tax_cents"],
            total_cents=money["total_cents"],
            created_at="2026-01-01T00:00:00.000Z",
        )

    def _next_id(self) -> str:
        """Hand out a fresh, unique order id."""
        global _sequence
        _sequence += 1
        return f"ord_{_sequence}"

File: examples/orders-service/src/services/orderService.ts
import type { Order, OrderRequest } from "../domain/order.js";
import { validateOrderRequest } from "../validation/orderValidator.js";
import { PricingService } from "../pricing/pricingService.js";
import { OrderRepository } from "../repository/orderRepository.js";
import { EmailService } from "../notifications/emailService.js";

let sequence = 0;

/** The heart of the flow: takes a request and turns it into a stored, confirmed order. */
export class OrderService {
  constructor(
    private readonly pricing: PricingService,
    private readonly repository: OrderRepository,
    private readonly email: EmailService,
  ) {}

  /** Validate, price, store, and confirm a new order. */
  placeOrder(request: OrderRequest): Order {
    validateOrderRequest(request);
    const money = this.pricing.price(request);
    const order = this.assemble(request, money);
    this.repository.save(order);
    this.email.sendOrderConfirmation(order);
    return order;
  }

  /** Look up an order that was placed earlier. */
  getOrder(id: string): Order | undefined {
    return this.repository.findById(id);
  }

  /** Combine the request with its computed money fields into a final order. */
  private assemble(
    request: OrderRequest,
    money: Pick<Order, "subtotalCents" | "discountCents" | "taxCents" | "totalCents">,
  ): Order {
    return {
      id: this.nextId(),
      customerId: request.customerId,
      lines: request.lines,
      ...money,
      createdAt: this.timestamp(),
    };
  }

  /** Hand out a fresh, unique order id. */
  private nextId(): string {
    sequence += 1;
    return `ord_${sequence}`;
  }

  /** When this order was placed. */
  private timestamp(): string {
    return "2026-01-01T00:00:00.000Z";
  }
}
